In [ ]:
import numpy as np
import pandas as pd

In [ ]:
mass_shootings_usa = pd.read_csv("/content/drive/MyDrive/csv_files/mass_shootings.csv", sep=",", header= 0, names=['ID', 'Date', 'State', 'City', 'Address',
       'Murders', 'Injured', 'Operations'])

In [ ]:
mass_shootings_usa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ID          2000 non-null   int64  
 1   Date        2000 non-null   object 
 2   State       2000 non-null   object 
 3   City        2000 non-null   object 
 4   Address     1993 non-null   object 
 5   Murders     2000 non-null   int64  
 6   Injured     2000 non-null   int64  
 7   Operations  0 non-null      float64
dtypes: float64(1), int64(3), object(4)
memory usage: 125.1+ KB


In [ ]:
mass_shootings_usa

ID            Date           State                           City  \
0     2363589   July 19, 2022        Nebraska                        Lincoln   
1     2362912   July 18, 2022         Georgia                        Atlanta   
2     2362215   July 17, 2022         Indiana                    Beech Grove   
3     2361849   July 17, 2022        Michigan                        Detroit   
4     2362011   July 17, 2022         Indiana                      Greenwood   
...       ...             ...             ...                            ...   
1995  1369109  April 11, 2019      California                    Los Angeles   
1996  1369148  April 11, 2019         Arizona                        Phoenix   
1997  1367011   April 9, 2019        Missouri                    Kansas City   
1998  1365052   April 7, 2019  North Carolina  Winston Salem (Winston-salem)   
1999  1365077   April 7, 2019         Indiana                   Indianapolis   

                          Address  Murders  Injured  Operations  
0              2800 block of T St        0        4         NaN  
1           440 Thomasvillle Blvd        0        5         NaN  
2                  1100 S 9th Ave        1        3         NaN  
3                             NaN        0        4         NaN  
4                1251 US Hwy 31 N        4        2         NaN  
...                           ...      ...      ...         ...  
1995     E 103rd St and S Main St        1        4         NaN  
1996         7112 W Minnezona Ave        3        2         NaN  
1997            5339 Michigan Ave        1        3         NaN  
1998              515 N Cherry St        0        7         NaN  
1999  3600 block of Roosevelt Ave        2        2         NaN  

[2000 rows x 8 columns]

In [ ]:
#A column with NaN values
mass_shootings_usa.drop("Operations", axis=1, inplace=True)
mass_shootings_usa

ID            Date           State                           City  \
0     2363589   July 19, 2022        Nebraska                        Lincoln   
1     2362912   July 18, 2022         Georgia                        Atlanta   
2     2362215   July 17, 2022         Indiana                    Beech Grove   
3     2361849   July 17, 2022        Michigan                        Detroit   
4     2362011   July 17, 2022         Indiana                      Greenwood   
...       ...             ...             ...                            ...   
1995  1369109  April 11, 2019      California                    Los Angeles   
1996  1369148  April 11, 2019         Arizona                        Phoenix   
1997  1367011   April 9, 2019        Missouri                    Kansas City   
1998  1365052   April 7, 2019  North Carolina  Winston Salem (Winston-salem)   
1999  1365077   April 7, 2019         Indiana                   Indianapolis   

                          Address  Murders  Injured  
0              2800 block of T St        0        4  
1           440 Thomasvillle Blvd        0        5  
2                  1100 S 9th Ave        1        3  
3                             NaN        0        4  
4                1251 US Hwy 31 N        4        2  
...                           ...      ...      ...  
1995     E 103rd St and S Main St        1        4  
1996         7112 W Minnezona Ave        3        2  
1997            5339 Michigan Ave        1        3  
1998              515 N Cherry St        0        7  
1999  3600 block of Roosevelt Ave        2        2  

[2000 rows x 7 columns]

In [ ]:
#Find null data
mass_shootings_usa[mass_shootings_usa["Address"].isnull()] 

ID               Date           State             City Address  \
3     2361849      July 17, 2022        Michigan          Detroit     NaN   
94    2360288      June 11, 2022      New Mexico       Alamogordo     NaN   
159   2354757       May 13, 2022       Louisiana    Napoleonville     NaN   
181   2294515        May 3, 2022          Kansas  Cowley (county)     NaN   
567   2115336  September 5, 2021        Missouri            Wyatt     NaN   
624   2089557    August 10, 2021     Mississippi       Clarksdale     NaN   
1397  1725735       July 4, 2020  South Carolina         Barnwell     NaN   

      Murders  Injured  
3           0        4  
94          0        5  
159         0        5  
181         0        4  
567         0        4  
624         2        2  
1397        0        5

In [ ]:
mass_shootings_usa = mass_shootings_usa.fillna("Not_specified")
mass_shootings_usa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   ID       2000 non-null   int64 
 1   Date     2000 non-null   object
 2   State    2000 non-null   object
 3   City     2000 non-null   object
 4   Address  2000 non-null   object
 5   Murders  2000 non-null   int64 
 6   Injured  2000 non-null   int64 
dtypes: int64(3), object(4)
memory usage: 109.5+ KB


In [ ]:
#Duplicate_data
mass_shootings_usa.duplicated()
mass_shootings_usa.shape

(2000, 7)

In [ ]:
mass_shootings_usa

ID            Date           State                           City  \
0     2363589   July 19, 2022        Nebraska                        Lincoln   
1     2362912   July 18, 2022         Georgia                        Atlanta   
2     2362215   July 17, 2022         Indiana                    Beech Grove   
3     2361849   July 17, 2022        Michigan                        Detroit   
4     2362011   July 17, 2022         Indiana                      Greenwood   
...       ...             ...             ...                            ...   
1995  1369109  April 11, 2019      California                    Los Angeles   
1996  1369148  April 11, 2019         Arizona                        Phoenix   
1997  1367011   April 9, 2019        Missouri                    Kansas City   
1998  1365052   April 7, 2019  North Carolina  Winston Salem (Winston-salem)   
1999  1365077   April 7, 2019         Indiana                   Indianapolis   

                          Address  Murders  Injured  
0              2800 block of T St        0        4  
1           440 Thomasvillle Blvd        0        5  
2                  1100 S 9th Ave        1        3  
3                   Not_specified        0        4  
4                1251 US Hwy 31 N        4        2  
...                           ...      ...      ...  
1995     E 103rd St and S Main St        1        4  
1996         7112 W Minnezona Ave        3        2  
1997            5339 Michigan Ave        1        3  
1998              515 N Cherry St        0        7  
1999  3600 block of Roosevelt Ave        2        2  

[2000 rows x 7 columns]

In [ ]:
stadistic_data = mass_shootings_usa.describe()

In [ ]:
stadistic_data.iloc[:, 1:]

Murders      Injured
count  2000.000000  2000.000000
mean      0.978500     4.187000
std       1.497384     2.274779
min       0.000000     0.000000
25%       0.000000     3.000000
50%       1.000000     4.000000
75%       1.000000     5.000000
max      23.000000    29.000000

## Analysis (2019-2022)

In [ ]:
#Mass shooting bloodiest
bloodiest = mass_shootings_usa[mass_shootings_usa["Murders"] == 23]
bloodiest

ID            Date  State     City            Address  Murders  \
1831  1466705  August 3, 2019  Texas  El Paso  7101 Gateway Blvd       23   

      Injured  
1831       23

In [ ]:
#Mass shooting with more injured
more_injured = mass_shootings_usa[mass_shootings_usa["Injured"] == 29]
more_injured

ID          Date     State           City                 Address  \
43  2348384  July 4, 2022  Illinois  Highland Park  Central Ave and 2nd St   

    Murders  Injured  
43        8       29

In [ ]:
#State with more mass shootings
state = mass_shootings_usa["State"].value_counts()
state

Illinois                219
California              153
Texas                   147
Pennsylvania            103
Louisiana                99
New York                 98
Florida                  88
Ohio                     79
Michigan                 70
Maryland                 67
Georgia                  66
Missouri                 64
North Carolina           64
South Carolina           62
Alabama                  51
Tennessee                48
Virginia                 45
Mississippi              41
New Jersey               41
Indiana                  40
District of Columbia     35
Colorado                 35
Wisconsin                31
Minnesota                25
Arkansas                 24
Washington               22
Kentucky                 21
Arizona                  20
Massachusetts            19
Nevada                   14
Oregon                   14
Iowa                     14
Connecticut              14
Delaware                 11
New Mexico               10
Oklahoma            

In [ ]:
print(f"""Illionois is the state with more mass shootings during the period 2019-2022.
Quantity: {np.max(state, axis=0)} mass shootings""")

Illionois is the state with more mass shootings during the period 2019-2022.
Quantity: 219 mass shootings


In [ ]:
# City with more mass shooting
city = mass_shootings_usa["City"].value_counts()
city

Chicago          165
Philadelphia      82
Baltimore         46
Houston           42
Washington        35
                ... 
Lake Ozark         1
Levelland          1
Manchester         1
Walnut Creek       1
Moreno Valley      1
Name: City, Length: 646, dtype: int64

In [ ]:
print(f"""Chicago is the city with more mass shootings during this period.
Quantity: {np.max(city, axis=0)} mass shootings""")

Chicago is the city with more mass shootings during this period.
Quantity: 165 mass shootings


In [ ]:
state.shape[0]

48

There have been mass shootings in 48 states in USA during the period 2019-2022 ☹.

Data extra: In USA there are 50 states.

In [ ]:
zero_murders = mass_shootings_usa["Murders"] == 0
zero_murders = mass_shootings_usa[zero_murders].shape[0]

In [ ]:
percentage = zero_murders/mass_shootings_usa.shape[0]
print(f"""{percentage*100} % of mass shootings during the period 2019-2022 haven't been killed.""")

46.6 % of mass shootings during the period 2019-2022 haven't been killed.


## Report (2019-2022)

1) The mass shooting bloodiest was in El Paso, Texas with 23 murders.

2) The mass shooting with more injured was in Highland Park, Illionois with 29 victims.

3) State with more mass shootings during this period was Illinois. Quantity: 219 mass shootings.

4) Chicago was the city with more mass shootings during this period.
Quantity: 165 mass shootings.

5) There have been mass shootings in 48 states in USA ☹. 

6) 46.6 % of mass shootings during the period 2019-2022 haven't been killed.

## Analysis per year

In [ ]:
def df_year(year):
  """Data Frame with only data about the year selected"""
  condition = mass_shootings_usa["Date"].str.contains(year)
  data_frame = mass_shootings_usa[condition]
  return data_frame

In [ ]:
def more_killed(df, year):
  """Mass shooting bloodiest in that year"""
  df_bloodiest = df[df["Murders"] == np.max(df["Murders"], axis=0)]
  df_bloodiest = df_bloodiest.loc[:, ["State", "City", "Murders"]]
  df_bloodiest["Year"] = year
  return df_bloodiest

In [ ]:
def more_injured(df, year):
  """Mass shooting with more injured"""
  df_injured = df[df["Injured"] == np.max(df["Injured"], axis=0)]
  df_injured = df_injured.iloc[:, [2, 3, 6]]
  df_injured["Year"] = year
  return df_injured

In [ ]:
def total_victims(df, year):
  df_victims = df.loc[:, ["State", "City", "Murders", "Injured"]]
  df_victims["Total_victims"] = df_victims["Murders"] + df_victims["Injured"]
  df_victims["Year"] = year
  df_victims = df_victims[df_victims["Total_victims"] == np.max(df_victims["Total_victims"], axis=0)]
  return df_victims

In [ ]:
def city_mass_shooting(df):
  """City with more mass shooting"""
  city = df["City"].value_counts()
  return city

In [ ]:
def states_mass_shooting(df):
  """In how many states had mass shootings."""
  q_states = df["State"].value_counts()
  return q_states.shape[0]

In [ ]:
def months_mass_shootings(df):
  """In which month had more mass shootings"""
  months_list = [
          "January", "February", "March", "April", 
          "May", "June", "July", "August", 
          "September", "October", "Novemeber", "December"
]
  frequency = {}
  for m in months_list:
    df_month = df[df['Date'].str.contains(m)]
    frequency[m] = df_month.shape[0]
  max_frequency = max(frequency.values())
  months = []
  for k, v in frequency.items():
    if v == max_frequency:
      months.append(k)
  return months, max_frequency


2019

In [ ]:
df_2019 = df_year("2019")

In [ ]:
bloodiest_19 = more_killed(df_2019, 2019)
bloodiest_19

State     City  Murders  Year
1831  Texas  El Paso       23  2019

In [ ]:
df_injured_19 = more_injured(df_2019, 2019)
df_injured_19

State     City  Injured  Year
1795  Texas   Odessa       23  2019
1831  Texas  El Paso       23  2019

In [ ]:
df_victims_19 = total_victims(df_2019, 2019)
df_victims_19

State     City  Murders  Injured  Total_victims  Year
1831  Texas  El Paso       23       23             46  2019

In [ ]:
city_19 = city_mass_shooting(df_2019)
city_19

Chicago                         28
Baltimore                       16
Philadelphia                    13
Saint Louis                     11
Los Angeles                      7
                                ..
Marshall                         1
Round Lake (Round Lake Park)     1
Claflin                          1
Beaumont                         1
Phoenix                          1
Name: City, Length: 189, dtype: int64

In [ ]:
states_19 = states_mass_shooting(df_2019)
states_19

40

In [ ]:
month_2019 = months_mass_shootings(df_2019)
print(f"""Months: {month_2019[0]} | Q: {month_2019[1]} mass shootings""")

Months: ['June'] | Q: 53 mass shootings


2020

In [ ]:
df_2020 = df_year("2020")
df_2020

ID               Date          State          City  \
1048  1891461  December 31, 2020        Indiana  Indianapolis   
1049  1890946  December 31, 2020         Oregon  Happy Valley   
1050  1886974  December 26, 2020       Illinois      Rockford   
1051  1886913  December 26, 2020        Florida         Tampa   
1052  1887059  December 26, 2020  Massachusetts          Lynn   
...       ...                ...            ...           ...   
1653  1588308    January 4, 2020        Florida         Ocala   
1654  1586839    January 2, 2020       New York      Brooklyn   
1655  1585746    January 1, 2020           Ohio     Cleveland   
1656  1585394    January 1, 2020  West Virginia    Huntington   
1657  1585894    January 1, 2020       Missouri   Saint Louis   

                            Address  Murders  Injured  
1048           8215 Center Run Road        1        3  
1049    9200 block of SE Idleman Rd        0        4  
1050                4007 E State St        3        3  
1051        N 20th St and E 7th Ave        2        2  
1052   Washington St and Quincy Ter        1        5  
...                             ...      ...      ...  
1653        800 block of NW 6th Ter        1        3  
1654                   6120 3rd Ave        0        4  
1655           1437 St Clair Ave NE        0        4  
1656                   1113 4th Ave        0        7  
1657  3000 block of S Jefferson Ave        3        2  

[610 rows x 7 columns]

In [ ]:
bloodiest_20 = more_killed(df_2020, 2020)
bloodiest_20

State                City  Murders  Year
1219      California             Aguanga        7  2020
1498         Alabama  Valhermoso Springs        7  2020
1598  North Carolina             Moncure        7  2020

In [ ]:
df_injured_20 = more_injured(df_2020, 2020)
df_injured_20

State        City  Injured  Year
1298  District of Columbia  Washington       21  2020

In [ ]:
df_victims_20 = total_victims(df_2020, 2020)
df_victims_20

State        City  Murders  Injured  Total_victims  Year
1298  District of Columbia  Washington        1       21             22  2020

In [ ]:
city_20 = city_mass_shooting(df_2020)
city_20

Chicago                53
Philadelphia           31
Brooklyn               15
Houston                13
Baltimore              10
                       ..
Paducah                 1
Brockton                1
Birmingham (Ensley)     1
Beverly Hills           1
Huntington              1
Name: City, Length: 288, dtype: int64

In [ ]:
states_20 = states_mass_shooting(df_2020)
states_20

44

In [ ]:
month_2020 = months_mass_shootings(df_2020)
print(f"""Months: {month_2020[0]} | Q: {month_2020[1]} mass shootings""")

Months: ['June'] | Q: 94 mass shootings


2021

In [ ]:
df_2021 = df_year("2021")
df_2021

ID               Date         State             City  \
356   2201535  December 31, 2021      Maryland  Capitol Heights   
357   2201716  December 31, 2021   Mississippi         Gulfport   
358   2201216  December 31, 2021    California      Los Angeles   
359   2200968  December 30, 2021  Pennsylvania     Philadelphia   
360   2201052  December 30, 2021      Missouri       Kirksville   
...       ...                ...           ...              ...   
1043  1893744    January 3, 2021       Florida            Miami   
1044  1892034    January 1, 2021      Arkansas       Fort Smith   
1045  1892704    January 1, 2021         Texas         Amarillo   
1046  1892114    January 1, 2021          Iowa       Sioux City   
1047  1891720    January 1, 2021      Illinois        Galesburg   

                           Address  Murders  Injured  
356                       Cindy Ln        0        4  
357        1200 block of Lewis Ave        4        3  
358           10211 S. Avalon Blvd        0        6  
359   5100 block of Germantown Ave        0        7  
360       700 block of E Dodson St        3        1  
...                            ...      ...      ...  
1043             10525 NW 24th Ave        0        8  
1044               3211 Towson Ave        0        7  
1045                 2650 Dumas Dr        1        3  
1046              2637 S Walker St        1        4  
1047       1000 block of Monroe St        0        4  

[692 rows x 7 columns]

In [ ]:
bloodiest_21 = more_killed(df_2021, 2021)
bloodiest_21

State      City  Murders  Year
819  California  San Jose       10  2021
945    Colorado   Boulder       10  2021

In [ ]:
df_injured_21 = more_injured(df_2021, 2021)
df_injured_21

State     City  Injured  Year
812  Florida  Hialeah       20  2021

In [ ]:
df_victims_21 = total_victims(df_2021, 2021)
df_victims_21

State     City  Murders  Injured  Total_victims  Year
812  Florida  Hialeah        3       20             23  2021

In [ ]:
city_21 = city_mass_shooting(df_2021)
city_21

Chicago         60
Philadelphia    24
Houston         17
Washington      16
Baltimore       12
                ..
Huntington       1
Sharon Hill      1
Carbondale       1
Gulfport         1
Galesburg        1
Name: City, Length: 321, dtype: int64

In [ ]:
states_21 = states_mass_shooting(df_2021)
states_21

45

In [ ]:
month_2021 = months_mass_shootings(df_2021)
print(f"""Months: {month_2021[0]} | Q: {month_2021[1]} mass shootings""")

Months: ['July'] | Q: 86 mass shootings


2022

In [ ]:
df_2022 = df_year("2022")
df_2022

ID             Date           State         City  \
0    2363589    July 19, 2022        Nebraska      Lincoln   
1    2362912    July 18, 2022         Georgia      Atlanta   
2    2362215    July 17, 2022         Indiana  Beech Grove   
3    2361849    July 17, 2022        Michigan      Detroit   
4    2362011    July 17, 2022         Indiana    Greenwood   
..       ...              ...             ...          ...   
351  2201641  January 1, 2022       Wisconsin      Kenosha   
352  2202001  January 1, 2022        Illinois       Peoria   
353  2201876  January 1, 2022  South Carolina       Dillon   
354  2201848  January 1, 2022         Indiana   South Bend   
355  2201677  January 1, 2022        Colorado       Denver   

                       Address  Murders  Injured  
0           2800 block of T St        0        4  
1        440 Thomasvillle Blvd        0        5  
2               1100 S 9th Ave        1        3  
3                Not_specified        0        4  
4             1251 US Hwy 31 N        4        2  
..                         ...      ...      ...  
351               1708 52nd St        1        3  
352  2000 block of SW Adams St        2        2  
353                 3261 Hwy 9        0        5  
354   1500 block of S Miami St        0        4  
355              1919 Blake St        2        2  

[356 rows x 7 columns]

In [ ]:
bloodiest_22 = more_killed(df_2022, 2022)
bloodiest_22

State    City  Murders  Year
142  Texas  Uvalde       22  2022

In [ ]:
df_injured_22 = more_injured(df_2022, 2022)
df_injured_22

State           City  Injured  Year
43  Illinois  Highland Park       29  2022

In [ ]:
df_victims_22 = total_victims(df_2022, 2022)
df_victims_22

State    City  Murders  Injured  Total_victims  Year
142  Texas  Uvalde       22       17             39  2022

In [ ]:
city_22 = city_mass_shooting(df_2022)
city_22

Chicago         24
Philadelphia    14
Baltimore        8
Houston          7
San Antonio      6
                ..
Memphis          1
Anniston         1
Stanwood         1
Uvalde           1
South Bend       1
Name: City, Length: 208, dtype: int64

In [ ]:
states_22 = states_mass_shooting(df_2022)
states_22

37

In [ ]:
month_2022 = months_mass_shootings(df_2022)
print(f"""Months: {month_2022[0]} | Q: {month_2022[1]} mass shootings""")

Months: ['May', 'June'] | Q: 64 mass shootings


## Report results:

In [ ]:
#Mass shootings bloodiest:
df_bloodiest = pd.concat([bloodiest_19, bloodiest_20, bloodiest_21, bloodiest_22])
df_bloodiest

State                City  Murders  Year
1831           Texas             El Paso       23  2019
1219      California             Aguanga        7  2020
1498         Alabama  Valhermoso Springs        7  2020
1598  North Carolina             Moncure        7  2020
819       California            San Jose       10  2021
945         Colorado             Boulder       10  2021
142            Texas              Uvalde       22  2022

In [ ]:
df = df_bloodiest.loc[:, ["State", "Murders"]].sort_values("Murders", ascending=False)
df["Position"] = np.arange(1, df.shape[0]+1)
df[0:3]

State  Murders  Position
1831       Texas       23         1
142        Texas       22         2
819   California       10         3

The bloodiest mass shooting during the 2019-2022 period was in Texas, El Paso, with 23 murders in 2019. In addition, the state of Texas was the scene of the two bloodiest mass shootings of this period.

In [ ]:
#Mass shootings with more injured:
df_injured = pd.concat([df_injured_19, df_injured_20, df_injured_21, df_injured_22])
df_injured

State           City  Injured  Year
1795                 Texas         Odessa       23  2019
1831                 Texas        El Paso       23  2019
1298  District of Columbia     Washington       21  2020
812                Florida        Hialeah       20  2021
43                Illinois  Highland Park       29  2022

The mass shooting with the most injuries was in Highland Park, Illinois, with an amount of 29. It can also be observed that there have always been between 20 or more injured during these shootings.



In [ ]:
#In which state and city there were the most victims (Murders+Injured).
dict = {
    "State": ["Texas", "District of Columbia", "Florida", "Texas"],
    "City": ["El Paso", "Washington", "Hialeah", "Uvalde"],
    "Total_victims": [46, 22, 23, 39],
    "Year": [2019, 2020, 2021, 2022]
}
df_victims = pd.DataFrame(dict)
df_victims

State        City  Total_victims  Year
0                 Texas     El Paso             46  2019
1  District of Columbia  Washington             22  2020
2               Florida     Hialeah             23  2021
3                 Texas      Uvalde             39  2022

The state of Texas has been protagonist of the two mass shootings with more victims in 2019 and 2022.

In [ ]:
#Cities with more mass shootings:
dict = {
    "City": ["Chicago", "Chicago", "Chicago", "Chicago"],
    "Amount": [28, 53, 60, 24],
    "Year": [2019, 2020, 2021, 2022],
}
df_city = pd.DataFrame(dict)
sumatory = pd.DataFrame([["Total", df_city["Amount"].sum(), ""]], columns=["City", "Amount", "Year" ], index=[4])
df_city.append(sumatory)

City  Amount  Year
0  Chicago      28  2019
1  Chicago      53  2020
2  Chicago      60  2021
3  Chicago      24  2022
4    Total     165

Chicago has been the city with more mass shooting of each year ☹.

In total: 165 mass shootings during 2019-2022 period.

In [ ]:
#In how many states had mass shootings
dict = {
    "Amount States": [f"{states_19} / 50", f"{states_20} / 50", f"{states_21} / 50", f"{states_22} / 50"],
    "%": [f"{(states_19/50)*100}%", f"{(states_20/50)*100}%", f"{(states_21/50)*100}%", f"{(states_22/50)*100}%"],
    "Year": [2019, 2020, 2021, 2022],
}
df_state = pd.DataFrame(dict)
df_state

Amount States      %  Year
0       40 / 50  80.0%  2019
1       44 / 50  88.0%  2020
2       45 / 50  90.0%  2021
3       37 / 50  74.0%  2022

In each year there have been mass shootings in more than 70% of states in the United States.

In [ ]:
#In which months occurred more mass shootings
dict = {
    "Month(s)": ['June', 'June', 'July', 'May & June'],
    "Amount": [month_2019[1], month_2020[1], month_2021[1], month_2022[1]],
    "Year": [2019, 2020, 2021, 2022]
}
df_month = pd.DataFrame(dict)
df_month

Month(s)  Amount  Year
0        June      53  2019
1        June      94  2020
2        July      86  2021
3  May & June      64  2022

In each year during the second quarter have occurred the largest numbers of mass shootings.